Zend/zend_opcode.c

```diff
@@ -118,6 +118,9 @@ ZEND_API void zend_type_release(zend_type type, bool persistent) {
 						zend_string_release(ZEND_TYPE_NAME(*sublist_type));
 					}
 				} ZEND_TYPE_LIST_FOREACH_END();
+				if (!ZEND_TYPE_USES_ARENA(*list_type) && persistent) {
+					free(ZEND_TYPE_LIST(*list_type));
+				}
```

Comment:
The first commit of this PR triggered a crash in a test case when run with tracing JIT enabled (probably a double-free in `destroy_zend_class`, but I'm not 100% sure...), so I switched to only freeing persistent, non-arena-allocated types, which fixed the crash.

This does seem too much of an ad-hoc solution to me though. A cleaner solution might be to introduce a new  `zend_type_release_internal()` function that would only be called when destroying internal classes/arg_infos...

What do you think?

Reply:
Why not just:
```c
    if (!ZEND_TYPE_USES_ARENA(type)) {
      pefree(ZEND_TYPE_LIST(type), persistent);
    }
```
Like below?

Reply:
That's what I initially did in 36a0722, however it triggered a crash under opcache+JIT for [this test case](https://github.com/php/php-src/blob/master/Zend/tests/type_declarations/dnf_types/variance/invalid_invariance1.phpt).

From my understanding of this particular edge case, `destroy_zend_class` is called due to the invalid invariance after opcache has already moved the types from arena to persistent memory, so `ZEND_TYPE_USES_ARENA(list_type)` returns false and we somehow end-up with a double-free... At least that's the conclusion I came to in the limited time I had.

That's why I asked about adding a new specialized function for internal types, where we can be 100% sure that neither arena nor opcache (nor a weird mix of both) is involved.

Reply:
Is it just the JIT or does this happen with just opcache? I don't have the best setup to work on right now :-/

But considering what you said, having a specialized function makes sense (and is not harmful)

Reply:
> Is it just the JIT or does this happen with just opcache?

From what I remember, just enabling opcache was enough to trigger it.